45-50% performance improvement in MPPI controller using Eigen library for computation.

Ayush1285 opened the pull request:
## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | #4237 #3351  |
| Primary OS tested on | Ubuntu |
| Robotic platform tested on | NA, Optimizer Benchmark |
| Does this PR contain AI generated software? | No |

---

## Description of contribution in a few bullet points

* Replaced xtensor with Eigen library for computation. 
* Performance increased by 45-50% with footprint and critics enabled.
* Currently WIP

## Description of documentation updates required from your changes
---

## Future work that may be required in bullet points


#### For Maintainers: <!-- DO NOT EDIT OR REMOVE -->
- [ ] Check that any new parameters added are updated in docs.nav2.org
- [ ] Check that any significant change is added to the migration guide
- [ ] Check that any new features **OR** changes to existing behaviors are reflected in the tuning guide
- [ ] Check that any new functions have Doxygen added
- [ ] Check that any new features have test coverage
- [ ] Check that any new plugins is added to the plugins page
- [ ] If BT Node, Additionally: add to BT's XML index of nodes for groot, BT package's readme table, and BT library lists

SteveMacenski commented on nav2_mppi_controller/CMakeLists.txt:
```diff
@@ -14,9 +14,11 @@ set(XTENSOR_USE_XSIMD 1)
 find_package(ament_cmake REQUIRED)
 find_package(xsimd REQUIRED)
 find_package(xtensor REQUIRED)
```
Note, if removing xtensor, we should remove it completely so we don't find_package, link, or in package.xml :-) 

Maybe not to change right now while still migrating, but making the note for the next steps

Ayush1285 replied:
Yes, I will remove them once the complete migration is done.

SteveMacenski commented on nav2_mppi_controller/CMakeLists.txt:
```diff
@@ -91,7 +93,7 @@ add_library(mppi_critics SHARED
 )
 
-set(libraries mppi_controller mppi_critics)
+set(libraries mppi_controller mppi_critics) #
```
```suggestion
set(libraries mppi_controller mppi_critics)
```

SteveMacenski commented on nav2_mppi_controller/CMakeLists.txt:
```diff
@@ -100,7 +102,7 @@ foreach(lib IN LISTS libraries)
 endforeach()
 
-install(TARGETS mppi_controller mppi_critics
+install(TARGETS mppi_controller
```
Readd>

SteveMacenski commented on nav2_mppi_controller/CMakeLists.txt:
```diff
@@ -116,7 +118,7 @@ if(BUILD_TESTING)
   set(ament_cmake_copyright_FOUND TRUE)
   ament_lint_auto_find_test_dependencies()
   add_subdirectory(test)
-  # add_subdirectory(benchmark)
```
Readd?

SteveMacenski commented on nav2_mppi_controller/CMakeLists.txt:
```diff
@@ -71,7 +73,7 @@ add_library(mppi_controller SHARED
   src/optimizer.cpp
   src/critic_manager.cpp
-  src/trajectory_visualizer.cpp
+  # src/trajectory_visualizer.cpp
```
TODO :-)

SteveMacenski commented on nav2_mppi_controller/benchmark/CMakeLists.txt:
```diff
@@ -13,7 +13,7 @@ foreach(name IN LISTS BENCHMARK_NAMES)
     ${dependencies_pkgs}
   )
   target_link_libraries(${name}
-    mppi_controller mppi_critics benchmark
```
Readd

SteveMacenski commented on nav2_mppi_controller/benchmark/controller_benchmark.cpp:
```diff
@@ -122,8 +120,7 @@ static void BM_DiffDrivePointFootprint(benchmark::State & state)
   std::string motion_model = "DiffDrive";
-  std::vector<std::string> critics = {{"GoalCritic"}, {"GoalAngleCritic"}, {"ObstaclesCritic"},
-    {"PathAngleCritic"}, {"PathFollowCritic"}, {"PreferForwardCritic"}};
+  std::vector<std::string> critics = {};
```
Across file: readd

SteveMacenski commented on nav2_mppi_controller/benchmark/optimizer_benchmark.cpp:
```diff
@@ -49,11 +47,11 @@ void prepareAndRunBenchmark(
+  unsigned int path_points = 150u;
   int iteration_count = 2;
-  double lookahead_distance = 10.0;
+  double lookahead_distance = 40.0;
```
Change back please :-)

SteveMacenski commented on nav2_mppi_controller/include/nav2_mppi_controller/models/control_sequence.hpp:
```diff
@@ -40,15 +35,15 @@ struct Control
-  xt::xtensor<float, 1> vx;
-  xt::xtensor<float, 1> vy;
-  xt::xtensor<float, 1> wz;
+  Eigen::ArrayXf vx;
```
I believe there are some Array1f, Array2f, etc definitions that would be more optimal given we know the dimensions of the matrix. Across the board, we should use explicit sizing when we know it which allows the compiler to make better optimizations

Ayush1285 replied:
Agreed

Ayush1285 replied:
Array2f means that the Array is one-dimensional and its size is fixed, which is 2. But in our case, we don't know the size of the control seq beforehand, right?

SteveMacenski replied:
See https://github.com/ros-navigation/navigation2/pull/4621#discussion_r1744084152

SteveMacenski commented on nav2_mppi_controller/include/nav2_mppi_controller/motion_models.hpp:
```diff
@@ -71,42 +63,33 @@ class MotionModel
    */
   virtual void predict(models::State & state)
   {
-    // Previously completed via tensor views, but found to be 10x slower
```
Curious - did you try this and was it better in Eigen vs xtensor? 

Also, please readd, I left it commented out for the reason of having a reference for later use

Ayush1285 replied:
I removed it because the xtensor implementation was copying shifted tensor, which is not the case now after the introduction of acceleration constraints. Should keep it removed? Or add it back.

SteveMacenski commented on nav2_mppi_controller/include/nav2_mppi_controller/motion_models.hpp:
```diff
@@ -71,42 +63,33 @@ class MotionModel
+        float & cvx_curr = state.cvx(j, i - 1);
         cvx_curr = std::clamp(cvx_curr, vx_last + min_delta_vx, vx_last + max_delta_vx);
-        state.vx(i, j) = cvx_curr;
-        vx_last = cvx_curr;
```
Readd the storing of last values, this is a good optimization since it removes lookups in the array (which are constant time, but not free-time)

Ayush1285 replied:
I'm accessing Array in column-major fashion, so the last value is of the different sampled batch.
By default, arrays are stored in memory in column-major order and Eigen is [optimized for column-major storage](https://eigen.tuxfamily.org/dox/group__TopicStorageOrders.html#:~:text=The%20default%20in%20Eigen%20is,best%20with%20column%2Dmajor%20matrices.) hence accessing elements contiguously reduces the cache misses of the CPU. 
I've tested them on a 4000*500 array, and accessing in column-major order was at least 2.5-3 times faster.

Ayush1285 replied:
````
#include <iostream>
#include <math.h>
#include <chrono>
#include <Eigen/Dense>

int main()
{
  Eigen::ArrayXXf vx1(4000, 500);
  vx1.setRandom();
  Eigen::ArrayXXf cvx1(4000, 500);
  cvx1.setRandom();

  Eigen::ArrayXXf vx2(4000, 500);
  vx2.setRandom();
  Eigen::ArrayXXf cvx2(4000, 500);
  cvx2.setRandom();
  float min_delta_vx = -0.2f;
  float max_delta_vx = 0.2f;

  std::chrono::steady_clock::time_point begin = std::chrono::steady_clock::now();
  
  for (unsigned int i = 1; i != 500; i++) {
    for (unsigned int j = 0; j != 4000; j++) {
      float & vx_last = vx1(j, i - 1);
      float & cvx_curr = cvx1(j, i - 1);
      cvx_curr = std::clamp(cvx_curr, vx_last + min_delta_vx, vx_last + max_delta_vx);
      vx1(j, i) = cvx_curr;
    }
  }

  std::chrono::steady_clock::time_point now = std::chrono::steady_clock::now();
  std::cout << "Time taken for column-major operation = " << std::chrono::duration_cast<std::chrono::microseconds>(now- begin).count() << "[us]" << std::endl;

  begin = std::chrono::steady_clock::now();

  for (unsigned int i = 0; i != 4000; i++) {
    float vx_last = vx2(i, 0);
    for (unsigned int j = 1; j != 500; j++) {
      float & cvx_curr = cvx2(i, j - 1);
      cvx_curr = std::clamp(cvx_curr, vx_last + min_delta_vx, vx_last + max_delta_vx);
      vx2(i, j) = cvx_curr;
      vx_last = cvx_curr;
    }
  }

  now = std::chrono::steady_clock::now();
  std::cout << "Time taken for Row-major operation = " << std::chrono::duration_cast<std::chrono::microseconds>(now- begin).count() << "[us]" << std::endl;
}
````

SteveMacenski commented on nav2_mppi_controller/include/nav2_mppi_controller/motion_models.hpp:
```diff
@@ -161,11 +144,18 @@ class AckermannMotionModel : public MotionModel
+    const auto vx_ptr = control_sequence.vx.data();
+    auto wz_ptr = control_sequence.wz.data();
+    int steps = control_sequence.vx.size();
+    for(int i = 0; i < steps; i++)
```
Curious: if there's the ability to run a function over each element of an array, can that be used here to have the function do the clamping and applied to the array?

Ayush1285 replied:
Yes, there is "select" method and we have UnaryExpr methods also. But "select" method was slow compared to accessing using the for loop. I think SIMD instruction is not supported for clamping operation, hence overhead of Eigen templates is slowing down the operation. 
But, I'll confirm after comparing manual looping across raw pointers with Eigen::select and Eigen::UnaryExpr.

SteveMacenski replied:
Try using `fabs` instead so it doesn't use floats (which simd sucks at). Maybe its actually the `abs` rather than the `clamp` operation which should be fully templated - though some googling makes me think that `std::min(max, std::max(min, v));` is actually faster due to some of clamp's internal functions that we don't need for this case. You could try that here even without the `UnaryExpr` and possibly get a speed up

Ayush1285 replied:
You're right, `fabs` and using `min(upperlimit, max(lowerlimit))` is faster.

Ayush1285 replied:
yes, there are a few places. I'll replace them and check for improvement.

SteveMacenski commented on nav2_mppi_controller/include/nav2_mppi_controller/motion_models.hpp:
```diff
@@ -161,11 +144,18 @@ class AckermannMotionModel : public MotionModel
+    int steps = control_sequence.vx.size();
+    for(int i = 0; i < steps; i++)
+    {
+      float && wz_constrained = std::abs(*(vx_ptr + i) / min_turning_r_);
```
Why all `i+1`? You could start with `i=1` - though I think you still want to apply the first index, no? That's what we did before

SteveMacenski replied:
Sorry, this was a dyslexic moment, I read `i` as `1`. Carry on...

Performance Q: worth storing the pointer so we can do a `curr_vx_ptr++` operation to march to the next one from the current one instead of starting with `vx_ptr` (start) `+ i` (iterating to it). The caching and compiler might solve this for it, but worth a try to see if there's a performance difference.

SteveMacenski commented on nav2_mppi_controller/include/nav2_mppi_controller/optimizer.hpp:
```diff
@@ -194,16 +188,16 @@ class Optimizer
    */
   void integrateStateVelocities(
     models::Trajectories & trajectories,
-    const models::State & state) const;
```
Removing const?

SteveMacenski commented on nav2_mppi_controller/include/nav2_mppi_controller/tools/noise_generator.hpp:
```diff
@@ -84,6 +79,10 @@ class NoiseGenerator
    */
   void reset(mppi::models::OptimizerSettings & settings, bool is_holonomic);
 
+  inline static std::normal_distribution<float> ndistribution_vx;
```
Why inline statics? There should only be 1 instance of the noise generator

Ayush1285 replied:
Oops, by mistake. I'll correct it.

SteveMacenski commented on nav2_mppi_controller/include/nav2_mppi_controller/tools/noise_generator.hpp:
```diff
@@ -99,9 +98,11 @@ class NoiseGenerator
+  Eigen::ArrayXXf noises_vy_;
+  Eigen::ArrayXXf noises_wz_;
+
+  inline static std::default_random_engine generator_;
```
Ditto

SteveMacenski commented on nav2_mppi_controller/include/nav2_mppi_controller/tools/utils.hpp:
```diff
@@ -309,21 +298,22 @@ auto shortest_angular_distance(
 
-  const auto dx = data.path.x - traj_x;
-  const auto dy = data.path.y - traj_y;
+  const auto dx = (data.path.x.transpose()).replicate(traj_x.rows(), 1) - traj_x;
```
I need to validate this but probably OK?

SteveMacenski commented on nav2_mppi_controller/src/critics/constraint_critic.cpp:
```diff
@@ -50,40 +48,33 @@ void ConstraintCritic::score(CriticData & data)
+    int n_cols = data.state.vx.cols();
+    Eigen::ArrayXXf sgn(n_rows, n_cols);
+    auto vx_ptr = data.state.vx.data();
+    for(int i = 0; i != n_cols; i++)
```
Q - is this the best way to do this?

Ayush1285 replied:
I tried few options and `sgn = vx.unaryExpr([](const float x){ return copysignf(1.0f, x);});` seems to be fastest, then `sgn = vx/vx.abs()`. But there is not much difference between these two.

SteveMacenski commented on nav2_mppi_controller/src/optimizer.cpp:
```diff
@@ -97,6 +92,10 @@ void Optimizer::getParams()
   getParam(motion_model_name, "motion_model", std::string("DiffDrive"));
 
   s.constraints = s.base_constraints;
+  NoiseGenerator::ndistribution_vx = std::normal_distribution(0.0f, s.sampling_std.vx);
```
Why not do this in the calss initialization? This seems odd to be here

SteveMacenski reviewed (changes requested):
I did a pretty fast review, so far from complete on each detail, but a good starting point!

SteveMacenski commented:
Let me know here when you want me to take a look again! I'm quite excited for this work - even if for no reason than to move to Eigen + 10% performance boost, since Eigen's release and support is much more known than xtensor's

Ayush1285 commented:
> Let me know here when you want me to take a look again! I'm quite excited for this work - even if for no reason than to move to Eigen + 10% performance boost, since Eigen's release and support is much more known than xtensor's

Sure, I'm trying a few optimizations and will push changes once I'm done.

Ayush1285 commented:
> Let me know here when you want me to take a look again! I'm quite excited for this work - even if for no reason than to move to Eigen + 10% performance boost, since Eigen's release and support is much more known than xtensor's

I've completed the migration to Eigen. But we need to make sure if functionality wise everything is correct or not.

SteveMacenski commented on nav2_mppi_controller/CMakeLists.txt:
```diff
@@ -1,22 +1,12 @@
 find_package(ament_cmake REQUIRED)
-find_package(xsimd REQUIRED)
-find_package(xtensor REQUIRED)
+find_package(Eigen3 REQUIRED)
```
Are the various `check_cxx_compiler_flag` useful for Eigen?

Ayush1285 replied:
-mfma is necessary for fast floating point fused multiply-add operations. [Answer by One of the Eigen Core maintainer](https://stackoverflow.com/questions/24296152/how-to-speed-up-this-c-program-with-eigen-library-against-matlab#:~:text=Finally%2C%20you%20might%20also%20consider%20using%20the%20devel%20branch%20and%20enable%20AVX%20(e.g.%2C%20%2Dmavx)%20and%20FMA%20if%20your%20CPU%20does%20support%20FMA%20(e.g.%2C%20%2Dmfma). He has also mentioned about enabling OpenMP for multi-threading.

Ayush1285 replied:
One interesting find, I tried removing the -fast-math flag and it speeds up the performance by a big margin.
For 2000 * 50 size:
xtensor: 11.6 ms avg.
Eigen with fast-math flag: 8.9 ms
Eigen without fast-math flag: 7.5 ms

SteveMacenski replied:
> And it seems there is no harm in keeping ISA flags(SSE/AVX).

There actually is, ARM based processors have inconsistent support for them or versions, so it makes release difficult. If they don't impact performance positively, we should remove them and removes an entire vector of potential problems (which we're currently running into in https://github.com/ros-navigation/navigation2/issues/4380 and a few other places in the past)

I'd be curious about OpenMP's changes in performance! We could make that a build option if it helps and those that want to use it can!

https://stackoverflow.com/questions/56547557/basic-ways-to-speed-up-a-simple-eigen-program there are some remarks here on AVX/fast-math, mfma. Might be worth doing a bit of research on Eigen-specific compiler optimizations, it looks like a rich vein and what was good for xtensor might not be right for Eigen. 

For fast-math, it might be worth testing on a couple of CPUs if you have them to make sure (if not, I can also test on my side, I have a few on my benchtop). What experiment are you running to get that performance change and/or does it include all the critics?

Ayush1285 replied:
> If they're important, lets keep them but I had hoped this would be a good source of removal of problems with Eigen

On my CPU only -mfma and -O3 impacted performance positively(and removal of fast-math also). There was no impact of AVX and SSE flags.

Ayush1285 replied:
> There actually is, ARM based processors have inconsistent support for them or versions, so it makes release difficult.

Oh, I wasn't aware of this. Once merge conflicts and build errors are resolved, then we can test on multiple CPUs with different flags.

Ayush1285 replied:
Hey @SteveMacenski, after fixing all the tests and system tests, these are the timings that I'm getting on my machine. I'm using optimizer_benchmark.cpp to compare the timings.

batch_size = 2000;
time_steps = 56;
path_points = 50;
iteration_count = 2;
lookahead_distance = 20;
critics = {{"ConstraintCritic"}, {"CostCritic"}, {"GoalCritic"},
    {"GoalAngleCritic"}, {"PathAlignCritic"}, {"PathFollowCritic"}, {"PathAngleCritic"}, {"PreferForwardCritic"}};
![image](https://github.com/user-attachments/assets/ca691780-e874-462e-843e-9c36f95d776c)
The result looks good. Can you try the PR on your machine and check if similar gains are reproducible on your machine also?

SteveMacenski replied:
Wow. So you're seeing a ~47% increase in performance? That's incredible. What CPU did you test this against?

Ayush1285 replied:
> What CPU did you test this against?

It is Intel i7-8750H 12 cores.

SteveMacenski commented on nav2_mppi_controller/benchmark/optimizer_benchmark.cpp:
```diff
@@ -118,8 +116,7 @@ static void BM_Omni(benchmark::State & state)
   std::string motion_model = "Omni";
-  std::vector<std::string> critics = {{"GoalCritic"}, {"GoalAngleCritic"}, {"ObstaclesCritic"},
-    {"TwirlingCritic"}, {"PathFollowCritic"}, {"PreferForwardCritic"}};
+  std::vector<std::string> critics = {};
```
Readd

SteveMacenski commented on nav2_mppi_controller/include/nav2_mppi_controller/controller.hpp:
```diff
@@ -20,7 +20,7 @@
 #include "nav2_mppi_controller/tools/path_handler.hpp"
 #include "nav2_mppi_controller/optimizer.hpp"
-#include "nav2_mppi_controller/tools/trajectory_visualizer.hpp"
+//#include "nav2_mppi_controller/tools/trajectory_visualizer.hpp"
```
Readd + the object

SteveMacenski commented on nav2_mppi_controller/include/nav2_mppi_controller/critic_data.hpp:
```diff
@@ -47,7 +42,7 @@ struct CriticData
   const models::Path & path;
 
-  xt::xtensor<float, 1> & costs;
+  Eigen::ArrayXf & costs;
```
Did we want to try if using `1f` rather than the `X` for fixed size would help performance?

SteveMacenski replied:
https://eigen.tuxfamily.org/dox/group__arraytypedefs.html 

I think it should be `ArrayXf` for 1xN and `Array2Xf` (or `ArrayXf`) for 2xN

Ayush1285 replied:
Yes, that's true. But in this case, we don't know the length of `costs` that's why I've used `ArrayXf`. Similarly, for `state` and `trajectories` variables, both rows and columns are not known at the compile time, so we have to use `ArrayXXf` only.

SteveMacenski commented on nav2_mppi_controller/include/nav2_mppi_controller/optimizer.hpp:
```diff
@@ -56,7 +50,7 @@ namespace mppi
 class Optimizer
 {
-public:
+public:  
```
Lint

SteveMacenski replied:
There's a few linting issues, I'm sure CI picks up on it once the merge conflicts are handled, or using ament_cpplint and ament_uncrustify locally

SteveMacenski commented on nav2_mppi_controller/include/nav2_mppi_controller/tools/noise_generator.hpp:
```diff
@@ -99,9 +94,14 @@ class NoiseGenerator
+  Eigen::ArrayXXf noises_vy_;
+  Eigen::ArrayXXf noises_wz_;
+
+  static std::default_random_engine generator_;
```
Why are these static?

SteveMacenski commented on nav2_mppi_controller/include/nav2_mppi_controller/tools/utils.hpp:
```diff
@@ -276,8 +265,8 @@ inline bool withinPositionGoalTolerance(
 template<typename T>
 auto normalize_angles(const T & angles)
```
This function I noted some energy in playing with exactly how to formulate it since the flamegraphs showed me that this was where non-trivial amounts of time were spent. Did you experiment with a few ways and see that this was indeed the fastest? Is this a chokepoint in gdb?

SteveMacenski replied:
In general, have you looked at some flame graphs / gdb output to see where the main bottlenecks are that might be worth revisiting now with Eigen

Ayush1285 replied:
I have not profiled them using a profiler, but I have manually compared the timings of many Eigen methods to see which one is faster. However, there are some places where redundant copies might be happening. RollColumns method seems to be one.

SteveMacenski replied:
@Ayush1285 just to sanity check here again - did you test this function though and the method used is the fastest?

Ayush1285 replied:
Yes, I've compared unaryExpr and manual for looping, and there was no difference for an array size of 10 million.

SteveMacenski commented on nav2_mppi_controller/src/noise_generator.cpp:
```diff
@@ -68,9 +74,9 @@ void NoiseGenerator::setNoisedControls(
-  xt::noalias(state.cvx) = control_sequence.vx + noises_vx_;
-  xt::noalias(state.cvy) = control_sequence.vy + noises_vy_;
-  xt::noalias(state.cwz) = control_sequence.wz + noises_wz_;
+  state.cvx = noises_vx_.rowwise() + control_sequence.vx.transpose();
```
It is not sensible to store the noises in such a way that we don't need to rowwise for addition?

Ayush1285 replied:
noises are stored in the correct order. But in Eigen, if we directly add a vector in a matrix then it throws the EIGEN_ASSERTION errors due to mismatch in no. of cols and rows of operands. `.rowwise()` specifies to perform row-wise addition of a vector with matrix.

SteveMacenski commented on nav2_mppi_controller/src/noise_generator.cpp:
```diff
@@ -107,17 +113,10 @@ void NoiseGenerator::noiseThread()
-    xt::noalias(noises_vy_) = xt::random::randn<float>(
-      {s.batch_size, s.time_steps}, 0.0f,
-      s.sampling_std.vy);
+  noises_vx_ = Eigen::ArrayXXf::NullaryExpr(s.batch_size, s.time_steps, [&] () {return ndistribution_vx_(generator_);});
```
Does this perform a copy into the variable? Is there a way to do this without that (i.e. looping) and is that faster?

Ayush1285 replied:
If we assign NullaryExpr to an `auto` variable then it assigns the abstract expression. But in this case, it is being evaluated in place. But I'm not sure if it first evaluates the data in a temporary place and then copies it to the variable. 
I checked the implementation of Random() method provided by Eigen, and they're also using NullaryExpr [Refer to link](https://gitlab.com/libeigen/eigen/-/blob/master/Eigen/src/Core/Random.h?ref_type=heads#:~:text=%3Ctypename%20Derived%3E-,inline%20const%20typename%20DenseBase%3CDerived%3E%3A%3ARandomReturnType%20DenseBase%3CDerived%3E%3A%3ARandom(Index,%7D,-/**%20%5Creturns%20a%20random)

SteveMacenski replied:
I agree this look reasonable and correct :+1:  Did you try looping though and seeing if that was faster? Sometimes particular implementations are faster for particular instances vs what the core library does that works in an abstract sense (which the author might think that array could be 1M+ in size which doing something like this might be more sensible). I don't know if `NullaryExpr` is something that using OpenMP/TBB have parallelized execution for.

SteveMacenski replied:
Ditto here

Ayush1285 replied:
I've tried looping instead of NullaryExpr. Both are almost equal in timing, but NullaryExpr is a little bit faster. Using OpenMP for a 2000 * 50(or even 5000 * 100) array is slower than normal sequential for loop.

SteveMacenski commented on nav2_mppi_controller/src/optimizer.cpp:
```diff
@@ -217,22 +214,11 @@ void Optimizer::prepare(
-  xt::view(control_sequence_.wz, -1) =
-    xt::view(control_sequence_.wz, -2);
-
+  control_sequence_.vx = utils::rollColumns(control_sequence_.vx, 1);
```
Does the roll also handle setting the last element to the second to last element?

Ayush1285 replied:
No, I'll add that part. I'll have to re-implement the rollColumns method because it is not shifting the values in place. Compiler is moving the values of the temporary matrix created inside this function, and this is causing bad memory handling, I think.

SteveMacenski replied:
@Ayush1285 did you complete this item?

Ayush1285 replied:
Yes

SteveMacenski commented on nav2_mppi_controller/src/optimizer.cpp:
```diff
@@ -247,33 +233,33 @@ void Optimizer::generateNoisedTrajectories()
-  float vy_last = control_sequence_.vy(0);
-  float wz_last = control_sequence_.wz(0);
-  for (unsigned int i = 1; i != control_sequence_.vx.shape(0); i++) {
+  float & vx_min = s.constraints.vx_min;
```
A `float &` is actually larger than a `float` copy, since the reference is the same size as a pointer. In general, the copy is better, but I think you could just use `s.constraints.XYZ` inline below rather than reducing them into another variable altogether

SteveMacenski commented on nav2_mppi_controller/src/optimizer.cpp:
```diff
@@ -247,33 +233,33 @@ void Optimizer::generateNoisedTrajectories()
+  float & vy = s.constraints.vy;
+  float & wz = s.constraints.wz;
+
+  float && max_delta_vx = s.model_dt * s.constraints.ax_max;
```
`Float &&`?

SteveMacenski commented on nav2_mppi_controller/src/optimizer.cpp:
```diff
@@ -247,33 +233,33 @@ void Optimizer::generateNoisedTrajectories()
     vx_last = vx_curr;
 
     float & wz_curr = control_sequence_.wz(i);
-    wz_curr = std::clamp(wz_curr, wz_last - max_delta_wz, wz_last + max_delta_wz);
```
Worth having a util for clamp that implements this instead of putting inline for each location? It is prone to copy+paste errors

SteveMacenski replied:
I think this is still nice-to-have, but I won't block for it

SteveMacenski replied:
@Ayush1285 thoughts on adding this? it would make things quite a bit more readable and less error prone for future modifiations

Ayush1285 replied:
We can add a util function. What will it implement? Clamping for all x, y, and z values, or just wrapping std::min(..., std::max())?

SteveMacenski replied:
Just a wrapper around `min(max())`

SteveMacenski commented on nav2_mppi_controller/src/optimizer.cpp:
```diff
@@ -306,78 +292,107 @@ void Optimizer::propagateStateVelocitiesFromInitials(
-  auto traj_x = xt::view(trajectory, xt::all(), 0);
-  auto traj_y = xt::view(trajectory, xt::all(), 1);
-  auto traj_yaws = xt::view(trajectory, xt::all(), 2);
+  auto traj_x = trajectory.col(0);
```
Are these copies or references?

Ayush1285 replied:
As I understand, these variables are abstract expressions representing the block view of the Eigen Array and storing references to the original variable. If we do eval(), it will copy the column data to the new variable.
[reference](https://eigen.tuxfamily.org/dox/TopicPitfalls.html#:~:text=In%20this%20example,.)

SteveMacenski reviewed (changes requested):
I did a first look through (didn't analyze in detail the math in the critics, but higher level programming items first), but generally looks good to me with some details to answer!

SteveMacenski commented:
I think you may not have done the last rebase / pull correctly. I'm seeing changes in files that don't relate to this PR. It could be due to the conflict and Git's attempt to show me the diffs it can with that in place, but that would need to be resolved before I can review again unfortunately :(

Ayush1285 commented:
> I think you may not have done the last rebase / pull correctly. I'm seeing changes in files that don't relate to this PR. It could be due to the conflict and Git's attempt to show me the diffs it can with that in place, but that would need to be resolved before I can review again unfortunately :(
> 
> Its a 3k line PR with 169 files, vast majority are unrelated

I've corrected the bad rebase/merge. Lint errors are also fixed.

SteveMacenski commented on nav2_mppi_controller/include/nav2_mppi_controller/critic_data.hpp:
```diff
@@ -44,10 +39,10 @@ namespace mppi
 {
   const models::State & state;
-  const models::Trajectories & trajectories;
+  models::Trajectories & trajectories;
```
Why removed const-ness?

Ayush1285 replied:
In critic files(ex path align critic), I was passing non-const in the template argument of the Eigen::Map, hence it was expecting non-const trajectories. I have corrected it now.

SteveMacenski commented on nav2_mppi_controller/include/nav2_mppi_controller/optimizer.hpp:
```diff
@@ -202,8 +196,8 @@ class Optimizer
   void integrateStateVelocities(
-    xt::xtensor<float, 2> & trajectories,
-    const xt::xtensor<float, 2> & state) const;
+    Eigen::Array<float, -1, 3> & trajectories,
```
Is there a typedef for -1 and dynamic so that an onlooker knows this means dynamic size? I.e. `Eigen::Dynamic`

Ayush1285 replied:
We can use Eigen::Dynamic instead of -1, that would be more self-explanatory.

SteveMacenski commented on nav2_mppi_controller/test/CMakeLists.txt:
```diff
@@ -43,10 +41,7 @@ target_link_libraries(critics_tests
 if(${TEST_DEBUG_INFO})
   target_compile_definitions(critics_tests PUBLIC -DTEST_DEBUG_INFO)
-endif()
+endif()
```
EOF line

SteveMacenski commented on nav2_mppi_controller/test/CMakeLists.txt:
```diff
@@ -12,6 +12,7 @@ set(TEST_NAMES
   optimizer_unit_tests
 )
 
+
```
Remove extra line here

SteveMacenski commented:
Should I have this another review? I know you're actively working and don't want to pester you mid-dev cycle :-)

SteveMacenski commented on nav2_mppi_controller/src/noise_generator.cpp:
```diff
@@ -81,9 +82,9 @@ void NoiseGenerator::reset(mppi::models::OptimizerSettings & settings, bool is_h
-    xt::noalias(noises_vx_) = xt::zeros<float>({settings_.batch_size, settings_.time_steps});
-    xt::noalias(noises_vy_) = xt::zeros<float>({settings_.batch_size, settings_.time_steps});
-    xt::noalias(noises_wz_) = xt::zeros<float>({settings_.batch_size, settings_.time_steps});
+    noises_vx_ = Eigen::ArrayXXf::Zero(settings_.batch_size, settings_.time_steps);
```
Another example of where you can do `.zeros()`?

Ayush1285 commented:
> Should I have this another review? I know you're actively working and don't want to pester you mid-dev cycle :-)

 Actually, I am also busy in other deliverables. So, I'm not able to allocate much time. But I'll let you know in 3-4 days maybe, once I've finished fixing the test failures and other review comments.

SteveMacenski commented:
Please request a review or otherwise comment when you want me to take a look again :-)

Ayush1285 commented:
> Please request a review or otherwise comment when you want me to take a look again :-)

Yes, I've requested for the review. All unit tests are passing now, but system tests are still failing.

SteveMacenski commented:
Retriggering the job - that could just be flaky. I do want to see coverage results before I review so I can use that in my review process. So I'll review once that is uploaded on the job success. I don't think there's an action item for you (just waiting on the build).

SteveMacenski commented:
Reviewing in the next hour or two. I commented on 2 previous comments on `clamp` and if the rolling was implemented. There's a couple more open comments on optimizations, but nothing we need to do immediately (just once we agree on everything else for the final touches). I appreciate your effort and diligence here!

SteveMacenski commented on nav2_mppi_controller/include/nav2_mppi_controller/tools/utils.hpp:
```diff
@@ -276,8 +265,12 @@ inline bool withinPositionGoalTolerance(
 {
-  auto theta = xt::eval(xt::fmod(angles + M_PIF, 2.0f * M_PIF));
-  return xt::eval(xt::where(theta < 0.0f, theta + M_PIF, theta - M_PIF));
+  return (angles + M_PIF).unaryExpr([&](const float x){
```
```suggestion
  return (angles + M_PIF).unaryExpr([&](const float x) {
```

Seems like a linting issue - I'm surprised if CI doesn't pick up on that

SteveMacenski commented on nav2_mppi_controller/include/nav2_mppi_controller/tools/utils.hpp:
```diff
@@ -276,8 +265,12 @@ inline bool withinPositionGoalTolerance(
+             float remainder = std::fmod(x, 2.0f * M_PIF);
+             return remainder < 0.0f ? remainder + M_PIF : remainder - M_PIF;
+             });
+  // auto theta = angles - (M_PIF * ((angles + M_PIF_2) * (1.0f / M_PIF)).floor());
```
Is this comment still worthwhile / tested for speed?

Ayush1285 replied:
yes, I had tested it for the speed, was not better so removed it now.

SteveMacenski commented on nav2_mppi_controller/include/nav2_mppi_controller/tools/utils.hpp:
```diff
@@ -612,6 +608,76 @@ struct Pose2D
+ * @param direction direction in which Array will be shifted.
+ *     1 for shift in right direction and -1 for left direction.
+ */
+inline void shiftColumnsByOnePlace(Eigen::Ref<Eigen::ArrayXXf> e, int direction)
```
Can we get some unit test for these 3 functions as new?

SteveMacenski commented on nav2_mppi_controller/src/critics/goal_angle_critic.cpp:
```diff
@@ -41,16 +41,15 @@ void GoalAngleCritic::score(CriticData & data)
-      xt::mean(xt::fabs(utils::shortest_angular_distance(data.trajectories.yaws, goal_yaw)), {1}) *
-      weight_, power_);
+  if(power_ > 1u) {
+    data.costs += (((utils::shortest_angular_distance(data.trajectories.yaws, goal_yaw).abs()).
```
For all: is there a `fabs` option to not case into double?

Ayush1285 replied:
Eigen itself is handling it, they have defined overloaded abs() for both float and double. 
![Screenshot from 2024-10-17 01-59-42](https://github.com/user-attachments/assets/47a0e751-8d18-430c-bd6c-192312eb9d75)

SteveMacenski commented on nav2_mppi_controller/src/optimizer.cpp:
```diff
@@ -248,32 +237,37 @@ void Optimizer::applyControlSequenceConstraints()
-  for (unsigned int i = 1; i != control_sequence_.vx.shape(0); i++) {
+  float vx_last = std::min(
+    s.constraints.vx_max, std::max(control_sequence_.vx(0), s.constraints.vx_min));
+  float vy_last = std::min(
```
Put vy in the `if` statement so its not computed for the ackermann/diff drive use cases

SteveMacenski reviewed (changes requested):
I got through everything except for the critic functions (but reviewed its unit tests so given that those work properly and have good coverage, I do not expect to find any issues there)

Ayush1285 commented:
> Reviewing in the next hour or two. I commented on 2 previous comments on `clamp` and if the rolling was implemented. There's a couple more open comments on optimizations, but nothing we need to do immediately (just once we agree on everything else for the final touches). I appreciate your effort and diligence here!

I'll address those review comments, but I'm not sure why system tests are failing consistently. I've tried running it manually but there is some issue with my docker environment. Can you try running it on your machine if possible?

SteveMacenski commented:
I'll retrigger CI to see if it'll work on a retry and just a fluke. Admittedly this week I'm running around finishing up ROSCon / ROS-I presentations, flying, and all day meetings in Chicago, so I can't promise how much time I'll have to test, but I will try. Hopefully I can work testing and debugging on the plane.

Ayush1285 commented:
> Edit: I just looked at the logs, and this looks like a real failure, not just some CI quirk.

@SteveMacenski  Yes, after investigation, I found out that there is some problem with CostCritic. After some control loops, it is always returning `all_trajectories_collide` as True, which causes failure. If I remove the CostCritic plugin, then it works fine. I'll need to look further to pinpoint the exact issue. Also, there is no test for CostCritc in the critics_test.cpp file.

SteveMacenski commented:
Back from ROSCon / ROS-I meetings now. Reviewing the software / remaining open items today. I think next steps from there are for me to play with this on my machine as well and:
- do some sanity testing of the various critics to make sure they're still working as I expect 
- do some functional sanity checking on the math for the (1) optimizer, (2) motion models, and (3) constraint compliance on the output optimal trajectory. These are software changes I want to scrutinize a little more functionally after a code review for my own piece of mind 
- Performance benchmarking :-) 
- Ask another non-me party to review the software changes and the performance on their practically applied platform as a second verification 

Is there anything else you think needs to be done here or is planned that we should factor in? 

I'm noticing on the code coverage report that there are a couple of lines that are no longer being tested that the UX is saying were covered before in the Obstacle Critic and possibly Optimizer (?). Its just a few stray lines, but makes my eyebrows raise a little bit since I'm not sure how that's possible if the tests are unchanged and they appear to be things that should be triggering :shrug:

SteveMacenski commented on nav2_mppi_controller/test/critics_tests.cpp:
```diff
@@ -433,28 +431,28 @@ TEST(CriticTests, TwirlingCritic)
+  state.wz.row(0) = Eigen::ArrayXf::NullaryExpr(30, [&] () {return normal_dist(engine);});
   critic.score(data);
-  EXPECT_NEAR(costs(0), 3.3, 4e-1);  // (mean of noise with mu=0, sigma=0.5 * 10.0 weight
+  EXPECT_NEAR(costs(0), 2.581, 4e-1);  // (mean of noise with mu=0, sigma=0.5 * 10.0 weight
```
@Ayush1285

SteveMacenski commented on nav2_mppi_controller/test/utils_test.cpp:
```diff
@@ -446,3 +447,108 @@ TEST(UtilsTests, RemovePosesAfterPathInversionTest)
+  array_1d << 1, 2, 3, 4;
+  utils::shiftColumnsByOnePlace(array_1d, 1);
+  EXPECT_EQ(array_1d.size(), 4);
+  EXPECT_EQ(array_1d(1), 1);
```
Please test the value at index `0`

SteveMacenski commented on nav2_mppi_controller/test/utils_test.cpp:
```diff
@@ -446,3 +447,108 @@ TEST(UtilsTests, RemovePosesAfterPathInversionTest)
+  EXPECT_EQ(array_1d.size(), 4);
+  EXPECT_EQ(array_1d(0), 5);
+  EXPECT_EQ(array_1d(1), 2);
+  EXPECT_EQ(array_1d(2), 3);
```
Please test the value at index `3`

SteveMacenski commented on nav2_mppi_controller/test/utils_test.cpp:
```diff
@@ -446,3 +447,108 @@ TEST(UtilsTests, RemovePosesAfterPathInversionTest)
+  EXPECT_EQ(array_2d(2, 2), 10);
+  EXPECT_EQ(array_2d(0, 3), 3);
+  EXPECT_EQ(array_2d(1, 3), 7);
+  EXPECT_EQ(array_2d(2, 3), 11);
```
Please test the `0`th column

SteveMacenski commented on nav2_mppi_controller/test/utils_test.cpp:
```diff
@@ -446,3 +447,108 @@ TEST(UtilsTests, RemovePosesAfterPathInversionTest)
+  EXPECT_EQ(array_2d(2, 1), 10);
+  EXPECT_EQ(array_2d(0, 2), 3);
+  EXPECT_EQ(array_2d(1, 2), 7);
+  EXPECT_EQ(array_2d(2, 2), 11);
```
Please test the `3`rd column

SteveMacenski commented on nav2_mppi_controller/include/nav2_mppi_controller/tools/utils.hpp:
```diff
@@ -612,6 +606,80 @@ struct Pose2D
+    yaws_between_points_corrected[i] = yaws[i] < M_PIF_2 ?
+      yaw_between_point : angles::normalize_angle(yaw_between_point + M_PIF);
+  }
+  return yaws_between_points_corrected;
```
Can we remove this copy?

Ayush1285 replied:
Sorry, I didn't get it. Are you referring about copy during return?

SteveMacenski replied:
Yes

Ayush1285 replied:
Will it be copied though? I think it won't be copied due to the named return-value optimization by compiler, since Eigen::Array has a defined Move constructor.

SteveMacenski commented on nav2_mppi_controller/include/nav2_mppi_controller/tools/utils.hpp:
```diff
@@ -612,6 +606,80 @@ struct Pose2D
+      angles::normalize_angle(yaw_between_point - goal_yaw)) < M_PIF_2 ?
+      yaw_between_point : angles::normalize_angle(yaw_between_point + M_PIF);
+  }
+  return yaws_between_points_corrected;
```
Can we remove this copy?

SteveMacenski commented on nav2_mppi_controller/include/nav2_mppi_controller/tools/utils.hpp:
```diff
@@ -612,6 +606,80 @@ struct Pose2D
+ *    1D Eigen Array by 1 place.
+ * @param e Eigen Array
+ * @param direction direction in which Array will be shifted.
+ *     1 for shift in right direction and -1 for left direction.
```
This appears to only work with `1` and `-1`, so should we error check by throwing if a non-`1` or `-1` is given as an input? I could see someone making the mistake of trying to roll by 2+ in some direction and being surprised that it didn't work

SteveMacenski commented on nav2_mppi_controller/src/critics/constraint_critic.cpp:
```diff
@@ -50,40 +48,31 @@ void ConstraintCritic::score(CriticData & data)
-            xt::maximum(data.state.vx - max_vel_, 0.0f) +
-            xt::maximum(min_vel_ - data.state.vx, 0.0f))) *
-          data.model_dt, {1}, immediate) * weight_, power_);
+      data.costs += (((((data.state.vx - max_vel_).max(0.0f) + (min_vel_ - data.state.vx).
```
Question: We have this `power > 1u` thing because in xtensor, theres a difference in performance when even taking `N^1` even though it is just `N`. If Eigen doesn't have that issue, we can remove the annoying if/else in all the critics

Ayush1285 replied:
Ah, Eigen also has this issue. I tried benchmarking after removing if/else, timing dropped from 3.65 ms to 3.90 ms for batch_size = 2000 and time_steps = 50.

SteveMacenski replied:
Cool, OK

SteveMacenski commented on nav2_mppi_controller/src/critics/constraint_critic.cpp:
```diff
@@ -50,40 +48,31 @@ void ConstraintCritic::score(CriticData & data)
+    Eigen::ArrayXXf sgn(n_rows, n_cols);
+    sgn = vx.unaryExpr([](const float x){return copysignf(1.0f, x);});
+
+    auto vel_total = (data.state.vx.square() + data.state.vy.square()).sqrt() * sgn;
```
More an ergonomic thing, but I'd like the `sgn *` before the rest of the expression. It reads more naturally IMO

SteveMacenski commented on nav2_mppi_controller/src/critics/constraint_critic.cpp:
```diff
@@ -50,40 +48,31 @@ void ConstraintCritic::score(CriticData & data)
-            xt::maximum(vel_total - max_vel_, 0.0f) +
-            xt::maximum(min_vel_ - vel_total, 0.0f))) *
-          data.model_dt, {1}, immediate) * weight_, power_);
+      data.costs += ((std::move((vel_total - max_vel_).max(0.0f) + (min_vel_ - vel_total).
```
Why does this one have the `std::move` but the prior didn't? As well for the 3x right below this one

Ayush1285 replied:
Removed it now. I added it when I was testing a few things, but forgot to remove it.

SteveMacenski commented on nav2_mppi_controller/src/critics/path_align_critic.cpp:
```diff
@@ -98,17 +96,20 @@ void PathAlignCritic::score(CriticData & data)
-    data.trajectories.yaws, xt::all(),
-    xt::range(0, _, trajectory_point_step_));
-  const auto traj_sampled_size = T_x.shape(1);
+  const auto T_x = Eigen::Map<const Eigen::ArrayXXf, 0,
```
Can you help me a little on this one breaking down these striding lines. I could spend some time tracking down the APIs myself, but I think it might be better to learn it from you to validate this

SteveMacenski replied:
We want to visit every batch (rows) but skip every `Nth` point in a given sample (cols). 

So I understand all but the `eigen::stride` population. To my (probably incorrect?) reading, it is striding `trajectory_point_step_` rows, but we only want to do cols. We want all rows (batches) evaluated. The `strided_traj_rows * trajectory_point_step_;` is making me think its skipping batches too. 

For your example, it would be `5`, not `10` to do all the rows (not skipping every-other)

SteveMacenski commented on nav2_mppi_controller/src/critics/prefer_forward_critic.cpp:
```diff
@@ -32,21 +34,19 @@ void PreferForwardCritic::initialize()
-      xt::sum(
-        std::move(
-          xt::maximum(-data.state.vx, 0)) * data.model_dt, {1}, immediate) * weight_, power_);
+    data.costs += Eigen::pow(
```
Why is this one `Eigen::pow` while the others use `.pow()`?

SteveMacenski commented on nav2_mppi_controller/src/critics/velocity_deadband_critic.cpp:
```diff
@@ -40,59 +40,35 @@ void VelocityDeadbandCritic::initialize()
-      weight_,
-      power_);
+    data.costs += ((((fabs(deadband_velocities_[0]) - data.state.vx.abs()).max(0.0f) +
+      (fabs(deadband_velocities_[1]) - data.state.vy.abs()).max(0.0f) +
```
We don't do `vy` / `[1]` for this condition if not doing omnidirectional (as well below)

SteveMacenski commented on nav2_mppi_controller/src/critics/cost_critic.cpp:
```diff
@@ -122,24 +120,30 @@ void CostCritic::score(CriticData & data)
+  int strided_traj_rows = data.trajectories.x.rows();
+  int outer_stride = strided_traj_rows * trajectory_point_step_;
+
+  const auto traj_x = Eigen::Map<const Eigen::ArrayXXf, 0,
```
STEVE: Review this one the other striding Q is answered

SteveMacenski commented on nav2_mppi_controller/src/critics/obstacles_critic.cpp:
```diff
@@ -150,33 +154,30 @@ void ObstaclesCritic::score(CriticData & data)
-      (critical_weight_ * raw_cost) +
-      (repulsion_weight_ * repulsive_cost_normalized),
-      power_);
+    data.costs += Eigen::pow(
```
Another `Eigen::pow`

SteveMacenski commented on nav2_mppi_controller/src/critics/obstacles_critic.cpp:
```diff
@@ -120,26 +119,31 @@ void ObstaclesCritic::score(CriticData & data)
+
+  // Default layout in eigen is column-major, hence accessing elements
+  // in column-major fashion to utilize L1 cache as much as possible
+  for(unsigned int i = 0; i != traj_len; i++) {
```
I think this should be looked at again, I don't think this is actually a 1:1 reimplementation. The behavior with `raw_cost`, `traj_cost` dont appear to me to be the same

Ayush1285 replied:
Yes, It was not the same exactly. I tried column-major looping to gain some performance by contiguous buffer access. But I've reverted now to the original implementation as behaviour was not same.

SteveMacenski reviewed (changes requested):
I got to everything except for the main optimizer file, I ran out of time :(

SteveMacenski reviewed (commented):
Completed the full review

Ayush1285 commented on nav2_mppi_controller/src/optimizer.cpp:
```diff
@@ -248,32 +237,38 @@ void Optimizer::applyControlSequenceConstraints()
-  float vy_last = control_sequence_.vy(0);
-  float wz_last = control_sequence_.wz(0);
-  for (unsigned int i = 1; i != control_sequence_.vx.shape(0); i++) {
+  float vx_last = std::min(
```
We are enforcing both velocity and acceleration constraints here in the loop. The acceleration constraint is enforced by checking the difference between the current and previous velocity. That won't apply to the 0th element.

SteveMacenski commented:
To prevent us from scrolling up through old comments before the last review: the only one left open is the Clamp question where I responded "yes, just `min(max())` to make it more ergonomic and help reduce likely issues with later modifications from other users"

SteveMacenski commented on nav2_mppi_controller/src/critics/obstacles_critic.cpp:
```diff
@@ -158,25 +159,21 @@ void ObstaclesCritic::score(CriticData & data)
-      (repulsion_weight_ * repulsive_cost_normalized),
-      power_);
+    data.costs +=
+      (critical_weight_ * raw_cost) + (repulsion_weight_ * repulsive_cost_normalized).pow(power_);
```
```suggestion
      ((critical_weight_ * raw_cost) + (repulsion_weight_ * repulsive_cost_normalized)).pow(power_);
```
I think this missed an important parenthesis, we should apply pow to both terms

SteveMacenski commented on nav2_mppi_controller/src/critics/obstacles_critic.cpp:
```diff
@@ -158,25 +159,21 @@ void ObstaclesCritic::score(CriticData & data)
       if (!near_goal) {
         repulsive_cost[i] += inflation_radius_ - dist_to_obj;
       }
-    }
```
I think there's a mistake here introduced. I think the `}` should be here and the `trajectory_collide` logic back under the parent scope

SteveMacenski reviewed (changes requested):
Otherwise, my only open other 2 things are the clamping and reviewing the test coverage 

Everything else is on me to do some functional testing for external validation and this is finally good to be merged! I'm very optimistic we can get this in before the end of the year!

SteveMacenski commented on nav2_mppi_controller/include/nav2_mppi_controller/motion_models.hpp:
```diff
@@ -71,42 +64,34 @@ class MotionModel
+      for (unsigned int j = 0; j != n_rows; j++) {
+        float vx_last = state.vx(j, i - 1);
+        float & cvx_curr = state.cvx(j, i - 1);
+        cvx_curr = std::min(vx_last + max_delta_vx, std::max(cvx_curr, vx_last + min_delta_vx));
```
min/max use the `clamp` util (and in all other locations we still do a min/max manually

SteveMacenski commented on nav2_mppi_controller/src/optimizer.cpp:
```diff
@@ -306,114 +298,130 @@ void Optimizer::propagateStateVelocitiesFromInitials(
-  auto yaw_sin = xt::roll(xt::eval(xt::sin(traj_yaws)), 1);
-  xt::view(yaw_cos, 0) = cosf(initial_yaw);
-  xt::view(yaw_sin, 0) = sinf(initial_yaw);
+  utils::shiftColumnsByOnePlace(traj_yaws, 1);
```
I believe the previous lines like `auto yaw_cos = xt::roll(xt::eval(xt::cos(traj_yaws)), 1);` will do the cosine of the `traj_yaws`, evaluate it into a tensor and roll that tensor in return. That would not then roll the `traj_yaws` itself. 

The lines that we have now look like they modify the `traj_yaws` which isn't correct since that is not a copy of `trajectory.col(2)` but a reference to it that can be modified, yes?

Ayush1285 replied:
Agreed, I missed this.

SteveMacenski commented on nav2_mppi_controller/src/optimizer.cpp:
```diff
@@ -306,114 +298,130 @@ void Optimizer::propagateStateVelocitiesFromInitials(
+  for(unsigned int i = 1; i != n_cols; i++) {
+    trajectories.yaws.col(i) = trajectories.yaws.col(i - 1) + state.wz.col(i) * settings_.model_dt;
+  }
+  utils::shiftColumnsByOnePlace(trajectories.yaws, 1);
```
Ditto

SteveMacenski reviewed (changes requested):
It continues to be odd to me that [we have 3 lines](https://app.codecov.io/gh/ros-navigation/navigation2/pull/4621) in the main optimizer that don't have codecov showing even though they're sequential operations with lines that are said to be covered. I assume its a compiler optimization artifact that these lines are optimized out, but it is curious to me. Nothing for you to worry about though unless this sparks a thought. 

```
+   auto yaw_sin = trajectories.yaws.sin().eval();
+   auto dy = (state.vx * yaw_sin).eval();
+   auto dy = (vx * yaw_sin).eval();
```

I retriggered CI, it should be green now. 

---

OK, next steps:

- [ ] Steve: do some sanity testing of the various things to make sure they're still working as I expect: (1) each critic, (2) optimizer, (3) motion models, and (4) constraint compliance on the output optimal trajectory. Run for a couple of hours.
- [x] Steve: Performance benchmarking :-)
- [ ] Steve: Ask another non-me party to review the software changes and the performance on their practically applied platform as a second verification 
- [ ] Steve: find ARM user to test

Anything else you feel should be done here - or are you happy with this as is? :smile: :partying_face:

SteveMacenski commented on nav2_mppi_controller/src/critics/cost_critic.cpp:
```diff
@@ -183,10 +187,10 @@ void CostCritic::score(CriticData & data)
   if (power_ > 1u) {
-    data.costs += xt::pow(
-      (std::move(repulsive_cost) * (weight_ / static_cast<float>(traj_len))), power_);
+    data.costs += (std::move(repulsive_cost) *
```
Remove the `std::move` I think we found

Ayush1285 commented:
> Anything else you feel should be done here

1. System tests are again failing consistently. In the GPS Waypoint Follower test, it passes the first waypoint, but after that action, the server refuses to accept the goal. Is it a generic failure? I've tested manually locally with multiple waypoints; it was okay there.
2. If we can test this on ARM platforms facing issues for xtensor, then it would be great to know if Eigen works there. Although we have removed all the SIMD extension flags, but as I know Eigen has explicit extrinsic written for SIMD extensions and then it is conditionally compiled based on whatever is supported by the processor. So, I think we might face problems here. Binaries generated by x86 might not work as desired on those ARM platforms.

SteveMacenski commented:
On ARM: I added that to my TODO list to find a deployed user, I have a couple in mind. I'll plan on doing benchmarking and the analysis steps in my first bullet this week

SteveMacenski reviewed (approved):
A migration guide entry on docs.nav2.org is necessary since this is such a big change. Add in the improvement metrics too!

Ayush1285 commented:
> A migration guide entry on docs.nav2.org is necessary since this is such a big change. Add in the improvement metrics too!

Yes, I was thinking that once you've done your testing and performance benchmarking, I'll add an entry on docs.nav2.org.

SteveMacenski commented:
Performance testing based on the Nav2 bringup default settings and differential drive model:
- Average for xtensor: 10.58ms
- Average for Eigen: 7.31ms
- Improvement: 30.9% 

High-level initial smoke testing looks good - no issues to report. I'm going to continue to dig into each of the critics just to prove to myself that they are working properly - but what I'm seeing from my testing, I'm not expecting to find anything.

SteveMacenski commented on nav2_mppi_controller/src/optimizer.cpp:
```diff
@@ -306,114 +298,135 @@ void Optimizer::propagateStateVelocitiesFromInitials(
+  size_t n_size = traj_yaws.size();
+
+  traj_yaws(0) = wz(0) * settings_.model_dt + initial_yaw;
+  float last_yaw = traj_yaws(0);
```
Would you agree this is the same and more concise?

```
  float last_yaw = initial_yaw;
  for(size_t i = 0; i != n_size; i++) {
    float & curr_yaw = traj_yaws(i);
    curr_yaw = last_yaw + wz(i) * settings_.model_dt;
    last_yaw = curr_yaw;
  }
```

If so, this and the X/Y entries below it could be updated as such :-) 

The other implementation this function could also be made similarly if the `last_val` variable is introduced

Ayush1285 replied:
Yes, this isthe  same. I'll update other instances of similar code.

SteveMacenski replied:
Sure thing :-)

Ayush1285 commented:
@SteveMacenski Opened the PR for docs changes : [Docs changes PR](https://github.com/ros-navigation/docs.nav2.org/pull/616)

Ayush1285 commented:
Hey @SteveMacenski, any updates on the beta testing? Do I need to handle the merge conflict?

SteveMacenski commented:
> any updates on the beta testing? 

I've been following up with folks, but with the holidays I'm receiving a pretty global "sorry for the delay, we'll get to it in mid-January" from folks. I'm pretty tempted to merge it given what I think we know, but I did want to get metrics from an ARM user before announcing this change since I think that is a critically key demographic for why this performance improvement is important. 

I will make a second wave for ARM users, if nothing else to just get some metrics on a Jetson or RPi. I do want to get this in before EoY though and I've seen nothing that would stop that from happening. 

> Do I need to handle the merge conflict?

Yes please :-)

Ayush1285 commented:
> I do want to get this in before EoY though and I've seen nothing that would stop that from happening.

Great, thanks!!.

> Yes please :-) Can't merge and harder for folks to test without it

Done.

SteveMacenski commented:
I have 1 report from an ARM user that the path align critic is having issues, but no other critics / issues in the general solution. I took a look at path align and don't see anything weird, but it might be worth giving that another look if you had 15-20 minutes to re-scrutinize it or any unique APIs we use there that may be having ARM issues. I don't see anything but the strides. 

To be noted, for the default bringup, the PathAlign critic and Cost critic are the only that use strides. 

The only thing I noted was that `int strided_traj_cols = data.trajectories.x.cols() / trajectory_point_step_ + 1;` --> I don't know how Eigen strides handle out of bounds or if it checks it, if we `+1` and that causes attempted access after the block. 

The other use in the cost critic makes sure this isn't possible with `int strided_traj_cols = floor(data.trajectories.x.cols() / trajectory_point_step_);`


>  Even at a low weight, enabling it caused erratic motion. When disabled, the performance was acceptable. Unfortunately, we had an issue with that robot today and I couldn't continue testing so we don't have any performance metrics yet. I will be on leave until January 6th, after which I’ll continue working on this. Keep you posted.

Ayush1285 commented:
> I have 1 report from an ARM user that the path align critic is having issues, but no other critics / issues in the general solution. I took a look at path align and don't see anything weird, but it might be worth giving that another look if you had 15-20 minutes to re-scrutinize it or any unique APIs we use there that may be having ARM issues. I don't see anything but the strides.
> 
> To be noted, for the default bringup, the PathAlign critic and Cost critic are the only that use strides.
> 
> The only thing I noted was that `int strided_traj_cols = data.trajectories.x.cols() / trajectory_point_step_ + 1;` --> I don't know how Eigen strides handle out of bounds or if it checks it, if we `+1` and that causes attempted access after the block.
> 
> The other use in the cost critic makes sure this isn't possible with `int strided_traj_cols = floor(data.trajectories.x.cols() / trajectory_point_step_);`
> 
> > Even at a low weight, enabling it caused erratic motion. When disabled, the performance was acceptable. Unfortunately, we had an issue with that robot today and I couldn't continue testing so we don't have any performance metrics yet. I will be on leave until January 6th, after which I’ll continue working on this. Keep you posted.

You're correct. Problem was in the calculation of `strided_traj_cols`. Since, first column in the trajectory will always be selected, the correct way should be: `strided_traj_cols = floor((data.trajectories.x.cols() - 1) / trajectory_point_step_) + 1;`.
Subtracting 1 for the first column and then adding it in the last.
For trajectories with even no. of columns, it was picking up one extra column with garbage values. It should be fixed now.

SteveMacenski commented:
Ah, glad we caught that! I just sent that update to the ARM user that was testing this -- lets see what they say (we're all getting back on the horse from the Holidays, so I imagine it'll be next week at this point)

rickarmstrong commented:
@Ayush1285 FYI I took a crack at building the `eigen_mppi` branch (`commit aab0b55497b6fe53fa4683a32d41379feffbd79b (HEAD -> eigen_mppi, origin/eigen_mppi`). Strangely, building `nav2_amcl` failed with 
```
Starting >>> nav2_amcl
--- stderr: nav2_amcl                                   
In file included from /root/tmp/nav2_ayushi_ws/src/navigation2/nav2_amcl/src/amcl_node.cpp:23:
/root/tmp/nav2_ayushi_ws/src/navigation2/nav2_amcl/include/nav2_amcl/amcl_node.hpp:32:10: fatal error: message_filters/subscriber.hpp: No such file or directory
   32 | #include "message_filters/subscriber.hpp"
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
This was in a Docker container created from image `osrf/ros jazzy-desktop-full   8ea079870e70   4 weeks ago   4.42GB`. Probably just something goofy in my setup, but figured I'd let you know anyhow.

ayush5ingh commented:
> @Ayush1285 FYI I took a crack at building the `eigen_mppi` branch (`commit aab0b55497b6fe53fa4683a32d41379feffbd79b (HEAD -> eigen_mppi, origin/eigen_mppi`). Strangely, building `nav2_amcl` failed with
> 
> ```
> Starting >>> nav2_amcl
> --- stderr: nav2_amcl                                   
> In file included from /root/tmp/nav2_ayushi_ws/src/navigation2/nav2_amcl/src/amcl_node.cpp:23:
> /root/tmp/nav2_ayushi_ws/src/navigation2/nav2_amcl/include/nav2_amcl/amcl_node.hpp:32:10: fatal error: message_filters/subscriber.hpp: No such file or directory
>    32 | #include "message_filters/subscriber.hpp"
>       |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> ```
> 
> This was in a Docker container created from image `osrf/ros jazzy-desktop-full 8ea079870e70 4 weeks ago 4.42GB`. Probably just something goofy in my setup, but figured I'd let you know anyhow.

I think it is due to because you're trying to build rolling branch of nav2 on Jazzy machine. More precisely, there are lot of C style headers, which were deprecated after launch of Jazzy. For example error that you are facing is due this PR: https://github.com/ros2/message_filters/pull/135
And I'm not sure if this change is back ported to Jazzy.

Ayush1285 commented:
> @Ayush1285 FYI I took a crack at building the `eigen_mppi` branch (`commit aab0b55497b6fe53fa4683a32d41379feffbd79b (HEAD -> eigen_mppi, origin/eigen_mppi`). Strangely, building `nav2_amcl` failed with
> 
> ```
> Starting >>> nav2_amcl
> --- stderr: nav2_amcl                                   
> In file included from /root/tmp/nav2_ayushi_ws/src/navigation2/nav2_amcl/src/amcl_node.cpp:23:
> /root/tmp/nav2_ayushi_ws/src/navigation2/nav2_amcl/include/nav2_amcl/amcl_node.hpp:32:10: fatal error: message_filters/subscriber.hpp: No such file or directory
>    32 | #include "message_filters/subscriber.hpp"
>       |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> ```
> 
> This was in a Docker container created from image `osrf/ros jazzy-desktop-full 8ea079870e70 4 weeks ago 4.42GB`. Probably just something goofy in my setup, but figured I'd let you know anyhow.

I think it is due to because you're trying to build rolling branch of nav2 on Jazzy machine. More precisely, there are lot of C style headers, which were deprecated after launch of Jazzy. For example error that you are facing is due this PR: https://github.com/ros2/message_filters/pull/135
And I'm not sure if this change is back ported to Jazzy.

mikeferguson commented:
> [](https://github.com/ros2/message_filters/pull/135)
> And I'm not sure if this change is back ported to Jazzy.

The tf2 headers are back ported to humble and jazzy - but I'm not sure they've sync'd yet to public debs (and even if the have, would have been in the last week or so and so you need to make sure to update your debs)

SteveMacenski commented:
From the ARM testing folks at Kiwibot:

> We tested the fix you mentioned on simulation and on a robot today and it seems to be working fine (including the PathAlignCritic), except for some oscillations that we think might require some tuning to fix. I will be working on that. Keep you posted.

It looks like that fix did the job. I'm helping them with some more tests and metrics gathering but its looking good :-)

jdgalviss commented:
Thanks for this PR! We (Kiwibot) tested it on one of our robots running ROS 2 Iron on Ubuntu 22.04 (Jetson AGX, ARM64). The compilation issues we previously encountered with xtensor are now resolved.

Using the default bringup parameters, we observed the following performance metrics:

17.1 ms average controller execution time
55–65% CPU usage on a single core during navigation
25% CPU usage on a single core when idle
0.5% overall memory usage
As @SteveMacenski mentioned, the controller’s performance has been good, and the issue with the PathAlignCritic is now fixed.

SteveMacenski commented:
I've heard enough good news at this point (only 2 weeks late from EOY)! 

@Ayush1285 please rebase / pull in main. If CI looks good, I'll merge!

Ayush1285 commented:
> I've heard enough good news at this point (only 2 weeks late from EOY)!
> 
> @Ayush1285 please rebase / pull in main. If CI looks good, I'll merge! Thanks so much for your tireless work and effort here, this is a big improvement in performance and long term maintainability. Even better, this enables this to work well and effectively on Jetson board which were not previously possible due to their relatively weak CPUs. The metrics that @jdgalviss reports from Kiwibot are a major improvement and make it possible to use MPPI even on those boards for production without using insane amount of resources or only able to run at sub-20hz.
> 
> I can't be happier with how this came out and I look forward to any other contributions you'd be interested in helping out with in Nav2! 😄

I've merged with the latest main. There are two un-related failures. One in collision monitor test: . 
`"[ERROR] [1736830701.557335092] [collision_monitor]: Error while getting parameters: parameter 'observation_sources' is not initialized`. Another one in docking server test: 
```[opennav_docking-1] [INFO] [1736830869.904053683] [docking_server]: Creating bond (docking_server) to lifecycle manager.
    [INFO] [lifecycle_manager-2]: sending signal 'SIGINT' to process[lifecycle_manager-2]``` . It seems flaky failures to me.

SteveMacenski commented:
Ugh, I'm sorry to do this to you, but it looks like there are merge conflicts from https://github.com/ros-navigation/navigation2/pull/4812 which changed the use of `path` to the actual goal `pose` to fix a bug in MPPI using the transformed path's end goal as the global goal and shortcutting some critics. 

Otherwise, I tested and this is good to go

Ayush1285 commented:
> Ugh, I'm sorry to do this to you, but it looks like there are merge conflicts from #4812 which changed the use of `path` to the actual goal `pose` to fix a bug in MPPI using the transformed path's end goal as the global goal and shortcutting some critics.

I've fixed the merge conflicts. Let's wait for the CI to run. The optimizer_benchmark was missing a change; it was not updated with the latest evalControl() API changes.

> This was a pretty big undertaking - what's your feeling about working on MPPI some more or switching gears into another part of the codebase / new algorithms?

I would like to contribute on other fronts. Playing with new control algorithms and improvising the costmap layer seems exciting ideas to me. Currently, I'm a little short on bandwidth but we can discuss/explore potential ideas for now.

SteveMacenski commented:
> I would like to contribute on other fronts. Playing with new control algorithms and improvising the costmap layer seems exciting ideas to me. Currently, I'm a little short on bandwidth but we can discuss/explore potential ideas for now.

Understood - want to chat a bit on the Nav2 Slack? By no means feel limited to those thoughts, the world's our oyster :laughing:

SteveMacenski commented:
@Ayush1285 only small thing that would make a nice followup PR: We talked about compressing some of the rollout stuff a bit more cleanly as so (as a rough hand-written-untested template)

```
  float last_yaw = initial_yaw;
  for(size_t i = 0; i != n_size; i++) {
    float & curr_yaw = traj_yaws(i);
    curr_yaw = last_yaw + wz(i) * settings_.model_dt;
    last_yaw = curr_yaw;
  }
```

Please ping me in Slack / email and lets chat about next projects when you have some bandwidth -- I can't thank you enough for this great work. I know this took alot of time and effort and it is well worth it for the global impact this has

YoungRicky7 commented:
Will this change merge into branch humble?

SteveMacenski commented:
No, this cannot be backported. It is a complete ABI/API breakage

V01DBREAKER commented:
Will this change be backported to Jazzy?

SteveMacenski commented:
No, it is a complete ABI rewrite. It is only available in Main, Kilted, and newer. Note that Main compiles against Jazzy if you want to, but won’t replace the build farm binaries.
